\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}


\usepackage{longtable,booktabs}
\usepackage{graphicx}
% grffile has become a legacy package: https://ctan.org/pkg/grffile
\IfFileExists{grffile.sty}{%
\usepackage{grffile}
}{}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\providecommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}

\RequirePackage[]{C:/Program Files/R/R-4.3.1/library/BiocStyle/resources/tex/Bioconductor}

\bioctitle[]{lute algorithm class definitions and uses}
    \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
\author{Sean K. Maden and Stephanie C. Hicks}
    \preauthor{\centering\large\emph}
  \postauthor{\par}
      \predate{\centering\large\emph}
  \postdate{\par}
    \date{17 January, 2024}

\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
% For Pandoc 2.8 to 2.11
\newenvironment{cslreferences}%
  {}%
  {\par}
% For pandoc 2.11+ using new --citeproc
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1 \everypar{\setlength{\hangindent}{\cslhangindent}}\ignorespaces\fi
  % set entry spacing
  \ifnum #2 > 0
  \setlength{\parskip}{#2\baselineskip}
  \fi
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}

% code highlighting
\definecolor{fgcolor}{rgb}{0.251, 0.251, 0.251}
\newcommand{\hlnum}[1]{\textcolor[rgb]{0.816,0.125,0.439}{#1}}%
\newcommand{\hlstr}[1]{\textcolor[rgb]{0.251,0.627,0.251}{#1}}%
\newcommand{\hlcom}[1]{\textcolor[rgb]{0.502,0.502,0.502}{\textit{#1}}}%
\newcommand{\hlopt}[1]{\textcolor[rgb]{0,0,0}{#1}}%
\newcommand{\hlstd}[1]{\textcolor[rgb]{0.251,0.251,0.251}{#1}}%
\newcommand{\hlkwa}[1]{\textcolor[rgb]{0.125,0.125,0.941}{#1}}%
\newcommand{\hlkwb}[1]{\textcolor[rgb]{0,0,0}{#1}}%
\newcommand{\hlkwc}[1]{\textcolor[rgb]{0.251,0.251,0.251}{#1}}%
\newcommand{\hlkwd}[1]{\textcolor[rgb]{0.878,0.439,0.125}{#1}}%
\let\hlipl\hlkwb
%
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
%
\newenvironment{Shaded}{\begin{myshaded}}{\end{myshaded}}
% set background for result chunks
\let\oldverbatim\verbatim
\renewenvironment{verbatim}{\color{codecolor}\begin{myshaded}\begin{oldverbatim}}{\end{oldverbatim}\end{myshaded}}
%
\newcommand{\KeywordTok}[1]{\hlkwd{#1}}
\newcommand{\DataTypeTok}[1]{\hlkwc{#1}}
\newcommand{\DecValTok}[1]{\hlnum{#1}}
\newcommand{\BaseNTok}[1]{\hlnum{#1}}
\newcommand{\FloatTok}[1]{\hlnum{#1}}
\newcommand{\ConstantTok}[1]{\hlnum{#1}}
\newcommand{\CharTok}[1]{\hlstr{#1}}
\newcommand{\SpecialCharTok}[1]{\hlstr{#1}}
\newcommand{\StringTok}[1]{\hlstr{#1}}
\newcommand{\VerbatimStringTok}[1]{\hlstr{#1}}
\newcommand{\SpecialStringTok}[1]{\hlstr{#1}}
\newcommand{\ImportTok}[1]{{#1}}
\newcommand{\CommentTok}[1]{\hlcom{#1}}
\newcommand{\DocumentationTok}[1]{\hlcom{#1}}
\newcommand{\AnnotationTok}[1]{\hlcom{#1}}
\newcommand{\CommentVarTok}[1]{\hlcom{#1}}
\newcommand{\OtherTok}[1]{{#1}}
\newcommand{\FunctionTok}[1]{\hlstd{#1}}
\newcommand{\VariableTok}[1]{\hlstd{#1}}
\newcommand{\ControlFlowTok}[1]{\hlkwd{#1}}
\newcommand{\OperatorTok}[1]{\hlopt{#1}}
\newcommand{\BuiltInTok}[1]{{#1}}
\newcommand{\ExtensionTok}[1]{{#1}}
\newcommand{\PreprocessorTok}[1]{\textit{#1}}
\newcommand{\AttributeTok}[1]{{#1}}
\newcommand{\RegionMarkerTok}[1]{{#1}}
\newcommand{\InformationTok}[1]{\textcolor{messagecolor}{#1}}
\newcommand{\WarningTok}[1]{\textcolor{warningcolor}{#1}}
\newcommand{\AlertTok}[1]{\textcolor{errorcolor}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor{errorcolor}{#1}}
\newcommand{\NormalTok}[1]{\hlstd{#1}}
%
\AtBeginDocument{\bibliographystyle{C:/Program Files/R/R-4.3.1/library/BiocStyle/resources/tex/unsrturl}}


\begin{document}
\maketitle

\packageVersion{lute 0.99.25}

\hypertarget{about}{%
\section{About}\label{about}}

This guide describes \texttt{lute}'s generics, methods, and classes for algorithms,
including deconvolution and marker selection algorithms. This may be useful to
algorithm developers and researchers interested in conducting systematic
algorithm benchmarks.

\hypertarget{background}{%
\subsection{Background}\label{background}}

The class structure used by \texttt{lute} is based on the \href{https://bioconductor.org/packages/release/bioc/html/bluster.html}{\texttt{bluster}} R/Bioconductor
package. It expands on that class structure by defining a hierarchy.

\hypertarget{motivation}{%
\subsection{Motivation}\label{motivation}}

Many algorithms are maintained and versioned in GitHub or Zenodo rather than a
routinely versioned repository such as Bioconductor or CRAN. This can prove an
obstacle when tracing package development and attempting comprehensive
benchmarks, as software that is not actively maintained can become deprecated
over time, and not all software will use compatible dependency versions
(Maden et al. (\protect\hyperlink{ref-maden_challenges_2023}{2023})).

\texttt{lute} classes can helo with this to (1.) encourage use of common Bioconductor
object classes (e.g.~\texttt{SummarizedExperiment}, \texttt{SingleCellExperiment},
\texttt{DelayedArray}, etc.) and (2.) to use more standard inputs and outputs to
encourage code reuse, discourage duplicated efforts, and enable more rapid and
exhaustive benchmarks.

\hypertarget{classes}{%
\section{Classes}\label{classes}}

In a general sense, the class hierarchy is a wrapper allowing access to many
algorithms using a single function and shared methods. However, it is possible
to share data reformatting and preprocessing tasks, making the hierarchy more
effectively similar to a workflow.

\hypertarget{typemarkerparam}{%
\subsection{typemarkerParam}\label{typemarkerparam}}

Topmost parameter class for cell type gene markers. This is used to manage the
marker IDs.

\hypertarget{deconvolutionparam}{%
\subsection{deconvolutionParam}\label{deconvolutionparam}}

This is the parent class for all deconvolution algorithm param objects. The
\texttt{deconvolutionParam} class is minimal, and simply defines slots for
\texttt{bulkExpression}, or a matrix of bulk expression data, and \texttt{returnInfo},
a logical value indicating whether the default algorithm output will be stored
and returned with standard output from running the \texttt{deconvolution()} method on
a valid algorithm param object.

\hypertarget{referencebasedparam}{%
\subsubsection{referencebasedParam}\label{referencebasedparam}}

As shown in the class hierarchy diagram (above), \texttt{referencebasedParam} is a
parent subclass inheriting attributes from \texttt{deconvolutionParam}. It is meant to
contain and manage all tasks shared by reference-based deconvolution algorithms,
or algorithms that utilize a cell type summary dataset. This is to be
distinguished from reference-free algorithms.

This param class adds slots for \texttt{referenceExpression}, the cell type reference
data, and \texttt{cellScaleFactors}, an optional vector of cell type size factors used
to transform the reference.

\hypertarget{independentbulkparam}{%
\subsubsection{independentbulkParam}\label{independentbulkparam}}

This class is a subset of \texttt{referencebasedParam} algorithms specifying explicit
samples used separately, such as for discrete training and test stages.

This param class adds a slot called \texttt{bulkExpressionIndependent}, which is for a
dataset of bulk samples independent from samples specified in the
\texttt{bulkExpression} slot.

\hypertarget{helper-functions}{%
\subsection{Helper functions}\label{helper-functions}}

\texttt{lute} provides a number of helper functions used to make the algorithm classes
work. These include the parent classes and subclasses, and several functions to
convert between object classes. These helper functions may be useful to
developers. The following table indicates the functions and a short summary of
what they do.

\begin{tabular}{c|c}
\hline
function\_name & description\\
\hline
`referenceFromSingleCellExperiment()` & Makes the Z cell atlas reference from a SingleCellExperiment.\\
\hline
`eset\_to\_sce()` & Convert ExpressionSet to SingleCellExperiment.\\
\hline
`sce\_to\_eset()` & Convert SingleCellExperiment to ExpressionSet\\
\hline
`se\_to\_eset()` & Convert SummarizedExperiment to ExpressionSet.\\
\hline
`get\_eset\_from\_matrix()` & Makes an ExpressionSet from a matrix.\\
\hline
`parseDeconvolutionPredictionsResults()` & Gets formatted predicted cell type proportions table from deconvolution results list.\\
\hline
`show()` & Method to inspect and summarize param object contents.\\
\hline
`deconvolution()` & Method to perform deconvolution with a param object.\\
\hline
`typemarkers()` & Method to get cell type markers with a param object.\\
\hline
`deconvolutionParam()` & Defines the principal parent class for all deconvolution method parameters.\\
\hline
`referencebasedParam()` & Class and methods for managing reference-based deconvolution methods.\\
\hline
`independentbulkParam()` & Class and methods for managing methods requiring independent bulk samples.\\
\hline
`typemarkersParam()` & Main constructor for class to manage mappings to the typemarkers() generic.\\
\hline
\end{tabular}

\hypertarget{algorithms}{%
\section{Algorithms}\label{algorithms}}

\hypertarget{findmarkers}{%
\subsection{findMarkers}\label{findmarkers}}

The param class \texttt{findmarkersParam} is defined for the function \texttt{findMarkers()}
from \texttt{scran} (see \texttt{?findmarkersParam}). This is a function to identify cell type
marker genes from a single-cell or single-nucleus expression dataset.

The \texttt{findmarkersParam} class is organized under its parent classes as
\texttt{typemarkersParam-\textgreater{}findMarkersParam}. It includes the \texttt{typemarkers()} method for
the identification of marker genes, and \texttt{show()} for inspecting the param
contents.

The following images annotate the constructor function and the \texttt{typemarkers()}
generic defined for the \texttt{findmarkersParam} class.

\hypertarget{nnls}{%
\subsection{NNLS}\label{nnls}}

The param class \texttt{nnlsParam} is defined for the function \texttt{nnls} from the \texttt{nnls}
R/CRAN package (see \texttt{?nnlsParam}). Non-negative least squares (NNLS) is commonly
used for deconvolution.

The \texttt{nnlsParam} class is organized under its parent classes as
\texttt{deconvolutionParam-\textgreater{}referencebasedParam-\textgreater{}nnlsParam}. It includes the
\texttt{deconvolution()} generic for cell type deconvolution, and the \texttt{show()} method
for inspecting the param contents.

The following images annotate the constructor function and the \texttt{deconvolution()}
generic defined for the \texttt{nnlsParam} class.

\hypertarget{bisque}{%
\subsection{Bisque}\label{bisque}}

The param class \texttt{bisqueParam} is defined for the function
\texttt{ReferenceBasedDeconvolution} from the \texttt{BisqueRNA} R/Bioconductor package (see
\texttt{?bisqueParam}). The Bisque algorithm adjusts on assay-specific biases arising
between the bulk and single-cell or single-nucleus platforms used to generate
expression datasets for deconvolution.

The \texttt{bisqueParam} class is organized under its parent classes as
\texttt{deconvolutionParam-\textgreater{}referencebasedParam-\textgreater{}independentbulkParam-\textgreater{}bisqueParam}.
It includes the \texttt{deconvolution()} generic for cell type deconvolution, and the
\texttt{show()} method for inspecting the param contents.

The following images annotate the constructor function and the \texttt{deconvolution()}
generic defined for the \texttt{bisqueParam} class.

\hypertarget{extensions}{%
\section{Extensions}\label{extensions}}

We demonstrated the extensibility and flexibility of \texttt{lute}'s generic, method,
and class system by extending support for additional algorithms beyond the 3
described above.

These algorithms can be used by sourcing the provided R/GitHub packages which
pair the classes and functions with YML files for easier dependency management.

\hypertarget{meanratios}{%
\subsection{meanRatios}\label{meanratios}}

The param class \texttt{meanratiosParam} is defined for the function
\texttt{get\_mean\_ratios2()} from the \texttt{DeconvoBuddies} R/GitHub package at \href{https://github.com/LieberInstitute/DeconvoBuddies}{LieberInstitute/DeconvoBuddies}.
This function uses the mean of cell type summary ratios to rank and select for
top marker genes.

The \texttt{meanratiosParam} class is organized under its parent classes as
\texttt{typemarkersParam-\textgreater{}meanratiosParam}. It includes the \texttt{typemarkers()} generic for
the identification of marker genes, and the \texttt{show()} method for inspecting the
param contents.

\texttt{meanratiosParam} is available from GitHub at \href{https://github.com/metamaden/meanratiosParam}{metamaden/meanratiosParam}.

\hypertarget{deconrnaseq}{%
\subsection{DeconRNASeq}\label{deconrnaseq}}

\texttt{deconvolutionParam-\textgreater{}referencebasedParam-\textgreater{}deconrnaseqParam}.

The param class \texttt{deconrnaseqParam} is defined for the function \texttt{DeconRNASeq}
(see \texttt{?deconrnaseqParam}) from the \texttt{DeconRNASeq} R/Bioconductor package
(\href{https://bioconductor.org/packages/release/bioc/html/DeconRNASeq.html}{link}).
The DeconRNASeq algorithm uses weighted averaged expression between types to
predicted cell type amounts more accurately for heterogeneous tissues
((\protect\hyperlink{ref-gong_deconrnaseq_2013}{\textbf{gong\_deconrnaseq\_2013?}})).

The \texttt{deconrnaseqParam} class is organized under its parent classes as
\texttt{deconvolutionParam-\textgreater{}referencebasedParam-\textgreater{}deconrnaseqParam}. It includes the
\texttt{deconvolution()} generic for cell type deconvolution, and the \texttt{show()} method
for inspecting the param contents.

The \texttt{deconrnaseqParam} class is available from GitHub at
\href{https://github.com/metamaden/deconrnaseqParam}{metamaden/deconrnaseqParam}

\hypertarget{epic}{%
\subsection{EPIC}\label{epic}}

The param class \texttt{epicParam} is defined for the function \texttt{EPIC} from the \texttt{EPIC}
R/GitHub package (see \texttt{?epicParam}). The EPIC algorithm was developed in blood
samples and incorporates cell size mRNA abundance (i.e.~cell size) and variance
normalizations (Racle and Gfeller (\protect\hyperlink{ref-racle_epic_2020}{2020})).

The \texttt{epicParam} class is organized under its parent classes as
\texttt{deconvolutionParam-\textgreater{}referencebasedParam-\textgreater{}epicParam}. It includes the
\texttt{deconvolution()} generic for cell type deconvolution, and the \texttt{show()} method for inspecting the param contents.

The \texttt{epicParam} class is available from GitHub at
\href{https://github.com/metamaden/epicParam}{metamaden/epicParam}

\hypertarget{music}{%
\subsection{MuSiC}\label{music}}

The param class \texttt{musicParam} is defined for the function
\texttt{ReferenceBasedDeconvolution} from the \texttt{MuSiC} R/GitHub package (see
\texttt{?musicParam}). The MuSiC algorithm adjusts on between-source variances for
reference data from multiple sources ((\protect\hyperlink{ref-wang_bulk_2019}{\textbf{wang\_bulk\_2019?}})).

The \texttt{musicParam} class is organized under its parent classes as
\texttt{deconvolutionParam-\textgreater{}referencebasedParam-\textgreater{}musicParam}. It includes the
\texttt{deconvolution()} generic for cell type deconvolution, and the \texttt{show()} method
for inspecting the param contents.

The \texttt{musicParam} class is available from GitHub at \href{https://github.com/metamaden/musicParam}{metamaden/musicParam}

\hypertarget{music2}{%
\subsection{MuSiC2}\label{music2}}

The param class \texttt{music2Param} is defined for 2 implementations of the \texttt{MuSiC2}
algorithm from the \texttt{MuSiC} and \texttt{MuSiC2} R/GitHub packages, respectively (see
\texttt{?music2Param}). The MuSiC2 algorithm pairs the features of the MuSiC algorithm
with an additional filter for marker genes differentially expressed between
cases and controls in the bulk and expression datasets ((\protect\hyperlink{ref-fan_music2_2022}{\textbf{fan\_music2\_2022?}})).

The \texttt{music2Param} class is organized under its parent classes as
\texttt{deconvolutionParam-\textgreater{}referencebasedParam-\textgreater{}independentbulkParam-\textgreater{}music2Param}.
It includes the \texttt{deconvolution()} generic for cell type deconvolution, and the
\texttt{show()} method for inspecting param contents.

The \texttt{music2Param} class is available from GitHub at \href{https://github.com/metamaden/music2Param}{metamaden/music2Param}

\hypertarget{conclusions}{%
\section{Conclusions}\label{conclusions}}

This vignette showed how \texttt{lute}'s classes and methods are extensible and
modular, and can encourage further development with standard algorithm I/O and
object class management. First, we described \texttt{lute}'s algorithm class hierarchy,
including how its parent classes and subclasses manage common tasks shared among
algorithms and a table of functions developers may find useful. Further, we
showed the annotated class and generic functions for algorithms supported by
\texttt{lute} out of the box. Finally, we detail additional algorithms supported by
R/GitHub packages that may be individually installed.

\hypertarget{session-info}{%
\section{Session info}\label{session-info}}

\begin{verbatim}
## R version 4.3.1 (2023-06-16 ucrt)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 11 x64 (build 22621)
## 
## Matrix products: default
## 
## 
## locale:
## [1] LC_COLLATE=English_United States.utf8 
## [2] LC_CTYPE=English_United States.utf8   
## [3] LC_MONETARY=English_United States.utf8
## [4] LC_NUMERIC=C                          
## [5] LC_TIME=English_United States.utf8    
## 
## time zone: America/Los_Angeles
## tzcode source: internal
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] lute_0.99.25                SingleCellExperiment_1.24.0
##  [3] SummarizedExperiment_1.32.0 Biobase_2.62.0             
##  [5] GenomicRanges_1.54.1        GenomeInfoDb_1.38.0        
##  [7] IRanges_2.36.0              S4Vectors_0.40.1           
##  [9] BiocGenerics_0.48.1         MatrixGenerics_1.14.0      
## [11] matrixStats_1.0.0           BiocStyle_2.30.0           
## 
## loaded via a namespace (and not attached):
##  [1] gtable_0.3.4              ggplot2_3.4.4            
##  [3] xfun_0.41                 lattice_0.22-5           
##  [5] vctrs_0.6.4               tools_4.3.1              
##  [7] bitops_1.0-7              generics_0.1.3           
##  [9] parallel_4.3.1            tibble_3.2.1             
## [11] fansi_1.0.5               cluster_2.1.4            
## [13] pkgconfig_2.0.3           BiocNeighbors_1.20.0     
## [15] Matrix_1.6-1.1            dqrng_0.3.1              
## [17] sparseMatrixStats_1.14.0  lifecycle_1.0.4          
## [19] GenomeInfoDbData_1.2.11   compiler_4.3.1           
## [21] munsell_0.5.0             statmod_1.5.0            
## [23] bluster_1.12.0            codetools_0.2-19         
## [25] htmltools_0.5.7           RCurl_1.98-1.13          
## [27] yaml_2.3.7                pillar_1.9.0             
## [29] crayon_1.5.2              BiocParallel_1.36.0      
## [31] limma_3.58.1              DelayedArray_0.28.0      
## [33] abind_1.4-5               metapod_1.10.0           
## [35] locfit_1.5-9.8            tidyselect_1.2.0         
## [37] rsvd_1.0.5                digest_0.6.33            
## [39] BiocSingular_1.18.0       dplyr_1.1.3              
## [41] bookdown_0.36             fastmap_1.1.1            
## [43] grid_4.3.1                colorspace_2.1-0         
## [45] cli_3.6.1                 SparseArray_1.2.0        
## [47] magrittr_2.0.3            S4Arrays_1.2.0           
## [49] utf8_1.2.4                edgeR_4.0.1              
## [51] scales_1.2.1              DelayedMatrixStats_1.24.0
## [53] rmarkdown_2.25            XVector_0.42.0           
## [55] igraph_1.5.1              scran_1.30.0             
## [57] ScaledMatrix_1.10.0       beachmat_2.18.0          
## [59] evaluate_0.23             knitr_1.45               
## [61] irlba_2.3.5.1             rlang_1.1.2              
## [63] Rcpp_1.0.11               scuttle_1.12.0           
## [65] glue_1.6.2                BiocManager_1.30.22      
## [67] rstudioapi_0.15.0         R6_2.5.1                 
## [69] zlibbioc_1.48.0
\end{verbatim}

\hypertarget{works-cited}{%
\section*{Works cited}\label{works-cited}}
\addcontentsline{toc}{section}{Works cited}

\hypertarget{refs}{}
\begin{CSLReferences}{1}{0}
\leavevmode\vadjust pre{\hypertarget{ref-maden_challenges_2023}{}}%
Maden, Sean K., Sang Ho Kwon, Louise A. Huuki-Myers, Leonardo Collado-Torres, Stephanie C. Hicks, and Kristen R. Maynard. 2023. {``Challenges and Opportunities to Computationally Deconvolve Heterogeneous Tissue with Varying Cell Sizes Using Single Cell {RNA}-Sequencing Datasets.''} arXiv. \url{https://doi.org/10.48550/arXiv.2305.06501}.

\leavevmode\vadjust pre{\hypertarget{ref-racle_epic_2020}{}}%
Racle, Julien, and David Gfeller. 2020. {``{EPIC}: {A} {Tool} to {Estimate} the {Proportions} of {Different} {Cell} {Types} from {Bulk} {Gene} {Expression} {Data}.''} Edited by Sebastian Boegel, Methods in {Molecular} {Biology}, 233--48. \url{https://doi.org/10.1007/978-1-0716-0327-7_17}.

\end{CSLReferences}


\end{document}
